### ISSUE TITLE
Allow managing flannel routes in an arbitrary routing table

### ISSUE THREAD

**Issue by tnyeanderson** [flannel-io/flannel#1767]
## Expected Behavior

Ideally, there would be a configuration option to set the table ID (and optional priority). When this option is set, the routes that get added for `flannel` on the host nodes would be added to the routing table with the corresponding ID. For example, if the config option is set to `123`, this might look like:

```
ip route add 10.42.0.0/24 via 10.42.0.0 dev flannel.1 onlink table 123
ip route add 10.42.1.0/24 dev cni0 proto kernel scope link src 10.42.1.1 table 123
ip route add 10.42.2.0/24 via 10.42.2.0 dev flannel.1 onlink table 123
```

Then, a rule would be added (with the priority, if provided) to direct flannel networks to use the new routing table first, then consult the main table for default gateway, etc. Using the `ip` tool:

```
ip rule add priority 1000 lookup 123
```

Provides the following rules list:

```
0:      from all lookup local
1000:   from all lookup 123
32766:  from all lookup main
32767:  from all lookup default
```

Additionally, a table alias (like `flannel`) might be added to `/etc/iproute2/rt_tables` or equivalent.

At this point, the normal behavior of flannel has been achieved, but this allows for more flexibility without duplicating routes. See the Context header below for my use case.

## Current Behavior

Currently all flannel routes are added to the `main` table, which is less flexible.

## Possible Solution

The `netlink` library [allows](https://github.com/vishvananda/netlink/blob/55c8b9515a0162749068d94e83bbb71911aa50c3/route.go#L63) setting `Table` on the route, so it should just be a matter of setting that at route creation/deletion on the host side.

## Context

Let's say I've created a `vpn0` interface on all my k8s nodes through some out-of-band process, and I want certain pods (based on some annotation) to have their internet traffic go through the `vpn0` interface instead of the default `eth0` using a CNI plugin. Based on the previous plugin in the chain (flannel), I know that the pod has an IP address of `10.42.1.3`. If I want to change the default route for packets coming from this IP, I need to use source-based routing. This requires setting up another table (`321`) with the `vpn0` default route, then adding a rule to direct that traffic using the new table. For example:

```
#############################
# Change the default route for the table to the VPN interface
#############################
ip route add default via <GATEWAY_IP> dev vpn0 table 321

#############################
# Evaluate routes with the new table for specific pods
#############################
ip rule add from 10.42.1.3/32 lookup 321
```

Unfortunately, now none of the flannel routes are present in the `321` table, so a lot of things start to fail until the flannel rules are duplicated in this table:

```
#############################
# Copy the flannel routes from the main table to the new table
#############################
ip route add 10.42.0.0/24 via 10.42.0.0 dev flannel.1 onlink table 321
ip route add 10.42.1.0/24 dev cni0 proto kernel scope link src 10.42.1.1 table 321
ip route add 10.42.2.0/24 via 10.42.2.0 dev flannel.1 onlink table 321
```

This would also need to be kept up to date as nodes are added/removed/etc. I don't want to have to duplicate these rules across tables. Chaining the rules won't work since there is a default route set.

With the proposed changes, the flannel routes might be stored in table `123`. Then I can create a rule with the correct priority (between `123` and `main`) and everything works... no need to duplicate the flannel routes!

```
0:      from all lookup local
1000:   from all lookup 123
2000:   from 10.42.1.3 lookup 321
32766:  from all lookup main
32767:  from all lookup default
```

**TLDR: This will make source-based routing or other policy-based routing much more simple because it would avoid duplicating the flannel routes in other tables.**

## Your Environment

Tested using k3s 1.26 on Ubuntu with the in-built flannel CNI.

**Comment by rbrtbnfgl** [comment]
Why do you need to replicate the flannel route is not enough to add the rule and the traffic from that IP will match that routing table?
I think that what you are proposing it's going out of scope of the CNI for your case maybe it will be better to use flannel with multus to manage the pods that you want to connect through the VPN.

**Comment by tnyeanderson** [comment]
>Why do you need to replicate the flannel route is not enough to add the rule and the traffic from that IP will match that routing table?

Because the pod needs to have those routes to be able to reach other pods and the outside world. Without the flannel routes, the pod can't reach anything. Adding the rule to tell the host that "pods with this IP should use this routing table" is not enough if that table doesn't include the flannel routes.

However, if the flannel routes are stored in their own table, they can be analyzed as part of the rules chain without issue. Also, since this is all regarding the default route (in my case) I can't just put the rule priority before the main table, because it will never get there. It will always try to use the default route that I set, which it can't do because it doesn't know how to talk through flannel anymore, unless I replicate the flannel rules in the new table... hence this feature request :)

>I think that what you are proposing it's going out of scope of the CNI for your case maybe it will be better to use flannel with multus to manage the pods that you want to connect through the VPN.

Multus would add interfaces and routing rules at the container level. What I am proposing affects routes at the host/node level. Unless I'm missing something, multus can't do what I'm wanting. I would love to be wrong about that!

In terms of being out of scope, the only change I'm proposing is to take the rules that flannel is *already creating* and allow me to define where those rules get created, allowing more flexibility for other CNI plugins, applications, or complex routing rules on the host. I personally don't think that is horribly out of scope.

This would be an additive configuration: if no table ID is set, just add them to the main table like always.

**Comment by rbrtbnfgl** [comment]
Thanks for your proposal. We are discussing if moving the route to a different routing table could affect other environment

**Comment by tnyeanderson** [comment]
@rbrtbnfgl Hi Roberto, any word on this?

**Comment by rbrtbnfgl** [comment]
Hi sorry for the late reply. We discussed about this but we have no plan to support this for now.

**Comment by tnyeanderson** [comment]
Thanks for following up! I did discover another way to do what I want (adjust default route for specific matchers):

```
# Create vpn0 interface
ip link add name vpn0 link eth0 type macvlan

# Set the vpn0 interface address
ip addr add 192.168.2.10/24 dev vpn0

# Bring the vpn0 interface up
ip link set dev vpn0 up

# Set up default route for interface in a separate table with ID 123
ip route add default via 192.168.2.1 dev vpn0 table 123

# NOTE: The priority for the rule which looks up the main table for every
# request is 32766, and each table should have a unique priority.

# First, check the main table for any non-default routes
ip rule add priority 32764 from all lookup main suppress_prefixlength 0

# Then, for traffic where the 5th bit mark is set (0x10 or 16) use the default gatway
# above defined in table 123.
ip rule add priority 32765 from all fwmark 0x10/0x10 lookup 123

# By default, rule 32766 looks up routes (including the default gateway) from
# the main table for all packets
```
